File: client/app/src/main/java/org/arghyam/puddle/data/repoImpl/LeaderboardRepoImpl.kt
package org.arghyam.puddle.data.repoImpl

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.get
import io.ktor.client.request.parameter
import io.ktor.client.request.url
import org.arghyam.puddle.BuildConfig
import org.arghyam.puddle.data.dto.responses.leaderboard.LeaderboardRes
import org.arghyam.puddle.domain.models.DataError
import org.arghyam.puddle.domain.models.Leaderboard
import org.arghyam.puddle.domain.models.Result
import org.arghyam.puddle.domain.repository.LeaderboardRepository

class LeaderboardRepoImpl(
    private val client: HttpClient
): LeaderboardRepository {
    override suspend fun getLeaderboard(userId: String): Result<Leaderboard, DataError> {
        return try {
            val leaderboardRes = client.get {
                url("${BuildConfig.SERVER_URL}/api/leaderboard/get-leaderboard")
                parameter("userId", userId)
            }.body<LeaderboardRes>()
            Result.Success(leaderboardRes.leaderboard)
        } catch (e: Exception) {
            e.printStackTrace()
            Result.Error(DataError.Network.SERVER_ERROR)
        }
    }
}

File: web/src/data/repoImpl/LeaderboardRepoImpl.ts
import axios from "axios";
import { LeaderboardRepository } from "../../domain/repository/LeaderboardRepository";
import { LEADERBOARD_ENDPOINT } from "../../utils/Constants";
import { GetLeaderboardRes } from "../models/leaderboard/GetLeaderboardRes";



export class LeaderboardRepoImpl implements LeaderboardRepository {
    async fetchLeaderboard(userId: string): Promise<GetLeaderboardRes> {
        try {
            const response = await axios.get<GetLeaderboardRes>(`${LEADERBOARD_ENDPOINT}/get-leaderboard?userId=${userId}`);
            return response.data;
          } catch (error) {
            throw error;
          }
    }
    
  
}
  